### PR TITLE
Tooltipdata and itembinding

### PR DESCRIPTION
**Pull request by steven-fuller**
Incorporates TooltipData and Soulbound/Unbound (BoE, BoU, BtA) categories.

### CODE REVIEW

File: core/constants.lua [hunk 1]
```diff
@@ -137,6 +137,25 @@ const.EXPANSION_TYPE = {
   LE_EXPANSION_DRAGONFLIGHT = 9,
 }
 
+---@enum TooltipDataItemBinding
+const.TOOLTIPDATA_ITEM_BINDING = {
```
Comment (Cidan) [inline core/constants.lua:141]: This should read TOOLTIP_DATA_ITEM_BINDING

File: core/constants.lua [hunk 2]
```diff
@@ -137,6 +137,25 @@ const.EXPANSION_TYPE = {
...
+  [Enum.TooltipDataItemBinding.Soulbound] = 3,
+  [Enum.TooltipDataItemBinding.BindToAccount] = 4,
+  [Enum.TooltipDataItemBinding.BindToBnetAccount] = 5,
+  [Enum.TooltipDataItemBinding.BindOnPickup] = 6,
+  [Enum.TooltipDataItemBinding.BindOnEquip] = 7,
+  [Enum.TooltipDataItemBinding.BindOnUse] = 8,
+}
+
+---@enum TooltipDataLineType
+const.TOOLTIPDATA_LINE_TYPE = {
```
Comment (Cidan) [inline core/constants.lua:154]: TOOLTIP_DATA_LINE_TYPE

File: core/constants.lua (hunk 1 above)
Comment (Cidan) [reply core/constants.lua:141]: Actually, looking at this again, do we have a particular reason we want to expose the fact this data originated from the tooltip? Why not ITEM_BINDING in this and all other enums?

File: data/items.lua [hunk 3]
```diff
@@ -182,6 +186,7 @@ function items:AttachItemInfo(data)
   local effectiveIlvl, isPreview, baseIlvl = GetDetailedItemLevelInfo(itemID)
   data.containerInfo = C_Container.GetContainerItemInfo(bagid, slotid)
   data.questInfo = C_Container.GetContainerItemQuestInfo(bagid, slotid)
+  data.tooltipData = C_TooltipInfo.GetBagItem(bagid, slotid)
```
Comment (Cidan) [inline data/items.lua:189]: Same reason/logic as above on partitioning this as tooltip data. Let's consider exposing this as just raw item data -- no need to tell the developer it came from the tooltip. I like indirection here, incase Blizzard adds an API down the line for this directly.

File: data/itemtooltips.lua [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  self:Update()
+end
+
+function itemTooltips:Update()
+ -- todo: Do we need to deal with TOOLTIP_DATA_UPDATE or does itemMixin:IsItemDataCached() gather tooltips?
+end
+
+---@param tooltipData tooltipData
+---@return Enum.TooltipDataItemBinding|nil
+function itemTooltips:GetItemBinding(tooltipData)
```
Comment (Cidan) [inline data/itemtooltips.lua:26]: While we're here, do we want to just extract everything we can from the tooltip, and add it to the data properties?

**Review by Cidan — CHANGES_REQUESTED** [review]
See comments, but general thoughts are:

* Remove TOOLTIP_DATA prefix
* Add binding info to the existing item properties -- the interface does not need to expose where the data came from, it just needs to expose the data. This is because I suspect Blizzard will add an API for some of this down the line, and representing the higher order data model as tooltip data would then be invalid.

File: core/constants.lua (hunk 1 above)
Comment (steven-fuller) [reply core/constants.lua:141]: Seemed to me like pulling in the Blizzard Enums was the right choice for maintainability
https://github.com/Gethe/wow-ui-source/blob/live/Interface/AddOns/Blizzard_APIDocumentationGenerated/TooltipInfoSharedDocumentation.lua.

File: data/itemtooltips.lua (hunk 4 above)
Comment (steven-fuller) [reply data/itemtooltips.lua:26]: This is exactly why I created a module for this. Binding is just one possibility. I'd have to ponder other use cases not covered by the item api

File: data/items.lua (hunk 3 above)
Comment (steven-fuller) [reply data/items.lua:189]: I had originally isolated all of the tooltip code in the plugin/module, but it was added to the items.lua data structure per your discord request - I had thought you wanted it there to facilitate the data caching - but I'm unclear on item vs tooltip data-caching mechanics tbh.
One reason would be that data.tooltipData is available to plugin developers, so they could avoid further scanning/loading.

> Add binding info to the existing item properties 

The 'binding' data I added is the extracted, but un-categorized version. I was thinking it might be useful to have the raw data available to plugins. (Like bi-sectional armor filters for transmogs)
`    binding = itemTooltips:GetItemBinding(data.tooltipData) --[[@as Enum.TooltipDataItemBinding]],
`

GetItemBinding could be generalized into items.lua instead of itemtooltips.lua by moving the const.BINDING_MAP out of frames and into the tooltip module - more like equipmentsets are handled. I'll whip this up this weekend.

Note that the current code only supports retail and compatibility routines would be needed for Classic where C_Tooltip isn't implemented

Comment (Cidan) [reply data/items.lua:189]: I don't think I was clear, sorry!

The reason I want to roll this into a data structure that isn't explicitly called out as tooltip is three fold:

1) I do actually believe that an API will come eventually. Blizzard in the last few years has been working to modernized a lot of the API's in general, and this is going to be an area of improvement for them.

2) I want to categorize data structures by their function, not by their source. For example, quest information comes from another API, but the item data structure doesn't mention that API, just that it's quest information.

3) Tooltip data needs further processing in all cases -- I want to provide this data already processed in the item data API, so that authors in the future don't have to reason about how to extract the data by making a call to itemtooltips.lua

I absolutely agree that itemtooltips.lua should remain -- this was a good idea and well thought out. The only change I'm asking for here is where the data lands in the `itemData` structure. Ideally, we would remove `data.tooltipData`, and only expose pre-calculated information for every item, i.e.

```lua
data.itemInfo.bound = SOME_ENUM
```

Does that make sense?

File: config/bags.lua [hunk 5]
```diff
@@ -98,6 +98,8 @@ function config:GetBagOptions(kind)
             values = {
               ["RecentItems"] = L:G("Recent Items"),
               ["Type"] = L:G("Type"),
+              ["Soulbound"] = L:G("Soulbound"),
+              ["Unbound"] = L:G("Unbound"),
```
Comment (Cidan) [inline config/bags.lua:102]: See comment below, but can we come up with a better label for Unbound? Unbound can apply to anything that isn't soulbound.

File: data/items.lua [hunk 6]
```diff
@@ -162,6 +166,39 @@ function items:ProcessBankContainer()
...
+    return itemTooltips:GetItemBinding(data.bagid, data.slotid) --[[Enum.ItemBinding|nil]]
+  end
+  if bindType == Enum.ItemBind.OnEquip then
+    return itemTooltips:GetItemBinding(data.bagid, data.slotid) --[[Enum.ItemBinding|nil]]
+  end
+  if bindType == Enum.ItemBind.OnUse then
+    return itemTooltips:GetItemBinding(data.bagid, data.slotid) --[[Enum.ItemBinding|nil]]
+  end
+
+  if bindType == Enum.ItemBind.Quest then -- we skip this as not useful / obsolete
```
Comment (Cidan) [inline data/items.lua:188]: You can just delete this line entirely, as it will fall through.

File: core/constants.lua [hunk 7]
```diff
@@ -137,25 +137,25 @@ const.EXPANSION_TYPE = {
...
-  [Enum.TooltipDataItemBinding.Account] = 1,
-  [Enum.TooltipDataItemBinding.BnetAccount] = 2,
-  [Enum.TooltipDataItemBinding.Soulbound] = 3,
-  [Enum.TooltipDataItemBinding.BindToAccount] = 4,
-  [Enum.TooltipDataItemBinding.BindToBnetAccount] = 5,
-  [Enum.TooltipDataItemBinding.BindOnPickup] = 6,
-  [Enum.TooltipDataItemBinding.BindOnEquip] = 7,
-  [Enum.TooltipDataItemBinding.BindOnUse] = 8,
+---@enum ItemBinding
+const.ITEM_BINDING = {
```
Comment (Cidan) [inline core/constants.lua:141]: Upper case these consts

ACCOUNT
SOUL_BOUND
BIND_ON_EQUIP
BIND_ON_USE

File: data/items.lua [hunk 8]
```diff
@@ -166,6 +166,39 @@ function items:ProcessBankContainer()
   end)
 end
 
+---@param data ItemData
+---@param bindType Enum.ItemBind
+---@return Enum.ItemBinding|nil
+function items:GetItemBinding(data, bindType)
```
Comment (Cidan) [inline data/items.lua:172]: This function does the same thing over and over again on every pattern match -- reduce this.

File: data/items.lua [hunk 9]
```diff
@@ -212,7 +244,7 @@ function items:AttachItemInfo(data)
     baseIlvl = baseIlvl --[[@as number]],
     itemIcon = C_Item.GetItemIconByID(itemID),
     isBound = C_Item.IsBound(itemLocation),
-    binding = itemTooltips:GetItemBinding(data.tooltipData) --[[@as Enum.TooltipDataItemBinding]],
+    binding = items:GetItemBinding(data,bindType) --[[Enum.ItemBinding|nil]],
```
Comment (Cidan) [inline data/items.lua:247]: Space after comma

File: data/itemtooltips.lua [hunk 10]
```diff
@@ -4,39 +4,91 @@ local addonName = ... ---@type string
...
+
 ---@class Events: AceModule
 local events = addon:GetModule('Events')
 
+---@class Debug: AceModule
+local debug = addon:GetModule('Debug')
+
+
+---@enum TooltipDataItemBinding
+const.TOOLTIP_DATA_ITEM_BINDING = {
```
Comment (Cidan) [inline data/itemtooltips.lua:20]: Consts must only be defined in constants.lua, please move these.

File: data/itemtooltips.lua [hunk 11]
```diff
@@ -4,39 +4,91 @@ local addonName = ... ---@type string
...
+  [Enum.TooltipDataItemBinding.Soulbound] = 3,
+  [Enum.TooltipDataItemBinding.BindToAccount] = 4,
+  [Enum.TooltipDataItemBinding.BindToBnetAccount] = 5,
+  [Enum.TooltipDataItemBinding.BindOnPickup] = 6,
+  [Enum.TooltipDataItemBinding.BindOnEquip] = 7,
+  [Enum.TooltipDataItemBinding.BindOnUse] = 8,
+}
+  
+---@enum TooltipDataLineType
+const.TOOLTIP_DATA_LINE_TYPE = {
```
Comment (Cidan) [inline data/itemtooltips.lua:33]: Please move.

File: data/itemtooltips.lua [hunk 12]
```diff
@@ -4,39 +4,91 @@ local addonName = ... ---@type string
...
+  [Enum.TooltipDataLineType.NestedBlock] = 19,
+  [Enum.TooltipDataLineType.ItemBinding] = 20,
+}
+
+---@class BindingMap
+---@type table<number, string>
+const.BINDING_MAP = {
```
Comment (Cidan) [inline data/itemtooltips.lua:40]: Please move.

File: data/items.lua (hunk 8 above)
Comment (Cidan) [reply data/items.lua:172]: It also looks like this, and the function GetItemBinding from tooltips convert from binding string to enum to string back and forth. Re-evaluate here and reduce to simpler terms if possible.

**Review by Cidan — COMMENTED** [review]
There will possibly be more changes to this in the next review, but let's start with this, since it's a lot already.

Thanks.

File: core/constants.lua [hunk 13]
```diff
@@ -371,12 +390,16 @@ const.DATABASE_DEFAULTS = {
...
+        Soulbound = true,
+        Unbound = true,
       },
       [const.BAG_KIND.BANK] = {
         Type = true,
         Expansion = false,
         TradeSkill = false,
         RecentItems = true,
+        Soulbound = true,
+        Unbound = true,
```
[the remark this answers is not in the document]
Comment (steven-fuller) [reply core/constants.lua:402]: 'Unbound' is pre-bound BoE/BoU and account bound currently. Exclusive with soulbound. I've thought about using 'Tradable', but that might be confusing as account bound (and warbound) items trade differently than BoE/BoU and I'm working up the code to allow a tradable BoP category for time limited tradable raid loot as its also in the tooltips only.

I'm very open to suggestions here to find a balance between too many options and confusing terminology.
Basically breaks down to items than can be put on the AH vs limited trades vs soulbound.

[the remark this answers is not in the document]
Comment (Cidan) [reply core/constants.lua:402]: How about we break out things that are on a bound timer as one of:

* Soulbound Tradable
* Soulbound (Timer)

WDYT?

File: config/bags.lua (hunk 5 above)
Comment (steven-fuller) [reply config/bags.lua:102]: See below. I think its worth distingushing 'never binding' items from pre-bound items.

File: data/items.lua [hunk 14]
```diff
@@ -198,7 +221,7 @@ function items:AttachItemInfo(data)
     sellPrice = sellPrice,
     classID = classID,
     subclassID = subclassID,
-    bindType = bindType,
+    bindType = bindType --[[@as Enum.ItemBind]],
```
Comment (steven-fuller) [inline data/items.lua:224]: GetItemInfo wiki and WoW API use out of date number or LE_ITEM_BIND instead of Enum.ItemBind
Maybe still true for classic?
